lib/ecto/repo/schema.ex

```diff
@@ -101,6 +108,35 @@ defmodule Ecto.Repo.Schema do
       {rows, header, placeholder_vals_list}
     end
   end
+  defp extract_header_and_fields(%Ecto.Query{select: %Ecto.Query.SelectExpr{expr: {:%{}, _ctx, args}}} = query, _schema, _dumper, _autogen_id, _placeholder_map, adapter) do
+    header = Enum.map(args, &elem(&1, 0))
+
+    query_and_params = Ecto.Adapter.Queryable.plan_query(:all, adapter, query)
+
+    {query_and_params, header, []}
+  end
+  defp extract_header_and_fields(%Ecto.Query{} = query, _schema, _dumper, _autogen_id, _placeholder_map, _adapter) do
+    raise ArgumentError, message: """
+    Cannot generate a fields list for insert_all from the given source query
```

Comment:
Nitpick: error messages start in lowercase and we can skip the message bit. :)

```suggestion
    raise ArgumentError, """
    cannot generate a fields list for insert_all from the given source query
```